### DISCUSSION TITLE
"compatible types" are not reflected as such in the API

### DISCUSSION THREAD

**Discussion by user1095108** [nlohmann/json#2394] [Q&A]
The only (perhaps not 100%) way to prevent exception 302 when assigning between json values is this very ineffective way:

```
if (!std::strcmp(a.type_name(), b.type_name()))
{
  then we can assign
}
```
Why not create a function are_compatible(a, b), or a member function is_compatible(a) with a bool result. In general, I'd like to see ways to avoid the exception 302 and compatible types seem to be key, but there is no way to discern whether 2 values are of compatible type.

**Comment by nlohmann** [comment]
I assume you do not use [`type()`](https://nlohmann.github.io/json/api/basic_json/type/) as you want to have all numbers treated as compatible?

**Reply by user1095108** [reply]
Yes, I wanted to enjoy maximum conversion capability between types - that's the whole idea behind **compatible types**, I suppose, but I can't assign a number to a bool, for example. For all I care, numbers could also be assignable to strings, like in js. So, this is a common problem I have, assigning numbers to bool values or vice versa and the same with strings. A string comparison seems an overkill to me for a concept that's mentioned in the docs. Compatible types, well let me check whether 2 types are compatible.

I don't think this string comparison gives a 100% answer. Sure, it works, but you may and probably will expand your notion of compatible types in the future, so I don't think this is the bulletproof solution I seek. And comparing type() results certainly isn't. That's checking for type equivalence, not compatibility.
